src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+};
+
+// distinguish between the types of PUBKEY
+// DEVICE_OWNED is specified by a path
+// THIRD_PARTY is specified by a pubkey hash
+enum {
+	PUBKEY_DEVICE_OWNED = 0,
+	PUBKEY_THIRD_PARTY = 1,
+};
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+bool expectsRequest()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->remainingScripts[ctx->level] > 0;
+}
+
+bool shouldFinish()
+{
+	// we finish if there are no more scripts to be processed
+	return ctx->level == 0 && ctx->remainingScripts[0] == 0;
+}
+
+static inline void scriptFinished()
+{
+	--ctx->remainingScripts[ctx->level];
+	TRACE("native script finished, level = %u, remainingScripts = %u", ctx->level, ctx->remainingScripts[ctx->level]);
+}
+
+static inline void complexScriptFinished()
+{
+	--ctx->level;
+	scriptFinished();
+}
+
+// UI
+
+static void deriveScriptHash_display_ui_callback()
+{
+	if (shouldFinish()) {
+		TRACE("whole native script received");
+		ui_idle();
+		// TODO: hashing, this is just a mock response
+		uint8_t buffer[28] = {0};
+		io_send_buf(SUCCESS, buffer, 28);
+	} else {
+		io_send_buf(SUCCESS, NULL, 0);
+	}
+}
+
+// Start complex native script
+
+void deriveNativeScriptHash_handleAll(read_view_t* view)
+{
+	// Parse data
+	ctx->remainingScripts[ctx->level] = parse_u4be(view);
+	ctx->type[ctx->level] = NATIVE_COMPLEX_SCRIPT_ALL;
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("level = %u, remaining scripts = %u", ctx->level, ctx->remainingScripts[ctx->level]);
+
+	// max possible length 26: "Confirm x scripts"
+	// where x is 2^32-1
+	char text[27];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d scripts", ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		// TODO: proper UI screen headers with numbered sections
+		"Script - ALL",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleAny(read_view_t* view)
+{
+	// Parse data
+	ctx->remainingScripts[ctx->level] = parse_u4be(view);
+	ctx->type[ctx->level] = NATIVE_COMPLEX_SCRIPT_ANY;
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("level = %u, remaining scripts = %u", ctx->level, ctx->remainingScripts[ctx->level]);
+
+	// max possible length 26: "Confirm x scripts"
+	// where x is 2^32-1
+	char text[27];
+	explicit_bzero(text, SIZEOF(text));
+	snprintf(text, SIZEOF(text), "Confirm %d scripts", ctx->remainingScripts[ctx->level]);
+	ASSERT(strlen(text) + 1 < SIZEOF(text));
+
+	ui_displayPaginatedText(
+		"Script - ANY",
+		text,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->nativeScript.nOfK.n = parse_u4be(view);
```

Comment:
Some explicit validation should be done before attempting to read the view (in other places, too).

Reply:
Isn't it enough that the `parse_u?be` functions do this validation before parsing?
```
		VALIDATE(view_remainingSize(view) >= width, ERR_NOT_ENOUGH_INPUT); \
```

Reply:
The convention in this code base is that whenever you are going to read from a view, you check that the view actually contains the data (check the buffer length). As I think about it now, however, it cannot be reasonably done for instance for bip44 paths (variable length, checked within the function), so it cannot be absolute. For security, it is better to check everything twice than risk not checking at all, but since we have problem with total code size, it seems preferable to drop checks that are obviously superfluous and just rely on the functions reading the view. You should create a separate PR for this: identify all situations like
```
		VALIDATE(view_remainingSize(view) >= 4, ERR_INVALID_DATA);
		params->protocolMagic = parse_u4be(view);

```
and remove the validation.